pkg/agent/run.go

```diff
@@ -381,66 +374,58 @@ func configureNode(ctx context.Context, nodeConfig *daemonconfig.Node, nodes typ
 	}
-	result = labels.Merge(nodeLabels, result)
-	return result, !equality.Semantic.DeepEqual(nodeLabels, result)
 }
 
-func updateLegacyAddressLabels(agentConfig *daemonconfig.Agent, nodeLabels map[string]string) (map[string]string, bool) {
-	ls := labels.Set(nodeLabels)
+func updateLegacyAddressLabels(agentConfig *daemonconfig.Agent, node *v1.Node, patch jsonpatch.PatchBuilder) {
+	ls := labels.Set(node.Labels)
+	patch = patch.WithPath("metadata", "labels")
 	if ls.Has(cp.InternalIPKey) || ls.Has(cp.HostnameKey) {
-		result := map[string]string{
-			cp.InternalIPKey: agentConfig.NodeIP,
-			cp.HostnameKey:   agentConfig.NodeName,
-		}
+		patch.AddIfNotEqual(ls, cp.InternalIPKey, agentConfig.NodeIP)
+		patch.AddIfNotEqual(ls, cp.HostnameKey, agentConfig.NodeName)
 
-		if agentConfig.NodeExternalIP != "" {
-			result[cp.ExternalIPKey] = agentConfig.NodeExternalIP
+		if agentConfig.NodeExternalIP == "" {
+			patch.RemoveIfHas(ls, cp.ExternalIPKey)
+		} else {
+			patch.AddIfNotEqual(ls, cp.ExternalIPKey, agentConfig.NodeExternalIP)
 		}
-
-		result = labels.Merge(nodeLabels, result)
-		return result, !equality.Semantic.DeepEqual(nodeLabels, result)
 	}
-	return nil, false
 }
 
 // updateAddressAnnotations updates the node annotations with important information about IP addresses of the node
-func updateAddressAnnotations(nodeConfig *daemonconfig.Node, nodeAnnotations map[string]string) (map[string]string, bool) {
+func updateAddressAnnotations(nodeConfig *daemonconfig.Node, node *v1.Node, patch jsonpatch.PatchBuilder) {
 	agentConfig := &nodeConfig.AgentConfig
-	result := map[string]string{
-		cp.InternalIPKey: util.JoinIPs(agentConfig.NodeIPs),
-		cp.HostnameKey:   agentConfig.NodeName,
-	}
-
-	if agentConfig.NodeExternalIP != "" {
-		result[cp.ExternalIPKey] = util.JoinIPs(agentConfig.NodeExternalIPs)
+	ls := labels.Set(node.Annotations)
+	patch = patch.WithPath("metadata", "annotations")
+
+	nodeIPs := util.JoinIPs(agentConfig.NodeIPs)
+	patch.AddIfNotEqual(ls, cp.InternalIPKey, nodeIPs)
+	patch.AddIfNotEqual(ls, cp.HostnameKey, agentConfig.NodeName)
+
+	if agentConfig.NodeExternalIP == "" {
+		patch.RemoveIfHas(ls, cp.ExternalIPKey)
+		patch.RemoveIfHas(ls, flannel.FlannelExternalIPv4Annotation)
+		patch.RemoveIfHas(ls, flannel.FlannelExternalIPv6Annotation)
+	} else {
+		nodeExternalIPs := util.JoinIPs(agentConfig.NodeExternalIPs)
+		patch.AddIfNotEqual(ls, cp.ExternalIPKey, nodeExternalIPs)
 		if nodeConfig.FlannelExternalIP {
 			for _, ipAddress := range agentConfig.NodeExternalIPs {
+				s := ipAddress.String()
 				if utilsnet.IsIPv4(ipAddress) {
-					result[flannel.FlannelExternalIPv4Annotation] = ipAddress.String()
-				}
-				if utilsnet.IsIPv6(ipAddress) {
-					result[flannel.FlannelExternalIPv6Annotation] = ipAddress.String()
+					patch.AddIfNotEqual(ls, flannel.FlannelExternalIPv4Annotation, s)
+				} else if utilsnet.IsIPv6(ipAddress) {
+					patch.AddIfNotEqual(ls, flannel.FlannelExternalIPv6Annotation, s)
```

Comment:
Why `AddIfNotEqual` instead of Add. If it is possible that there is already a label, shouldn't we call the `remove` action too?